### ISSUE TITLE
Uploading file doesn't work on weblogic 12.1.3 and spring boot 1.2.3.RELEASE

### ISSUE THREAD

**Issue by jeromedoucet** [spring-projects/spring-boot#2862]
The file upload sample here : https://spring.io/guides/gs/uploading-files/ works fine as a stand alone app but failed when deployed on weblogic 12.1.3

The response to the upload POST request is a 400 "exception":"org.springframework.web.bind.MissingServletRequestParameterException","message":"Required MultipartFile parameter 'file' is not present"

When debugging, it seems that not parts are return when calling getParts() on a the HttpServletRequest inside parseRequest(HttpServletRequest request) of the classStandardMultipartHttpServlet :

```java

private void parseRequest(HttpServletRequest request) {
		try {
			Collection<Part> parts = request.getParts(); // empty collection here
			this.multipartParameterNames = new LinkedHashSet<String>(parts.size());
			MultiValueMap<String, MultipartFile> files = new LinkedMultiValueMap<String, MultipartFile>(parts.size());
			for (Part part : parts) {
				String filename = extractFilename(part.getHeader(CONTENT_DISPOSITION));
				if (filename != null) {
					files.add(part.getName(), new StandardMultipartFile(part, filename));
				}
				else {
					this.multipartParameterNames.add(part.getName());
				}
			}
			setMultipartFiles(files);
		}
		catch (Exception ex) {
			throw new MultipartException("Could not parse multipart servlet request", ex);
		}
	}
```
After some search and experiments, we finally succed by downgrading spring-boot-autoconfigure module from 1.2.3 to 1.2.0

**Comment by wilkinsona** [comment]
@jeromedoucet Did you only downgrade the auto-configure module and leave all your other dependencies unchanged?

Not much has changed around multi-part auto configuration between 1.2.0 and 1.2.3: it's just the top three commits [here](https://github.com/spring-projects/spring-boot/commits/master/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/MultipartAutoConfiguration.java).

Can you share an app that illustrates the problem with Boot 1.2.3?

**Comment by mssoftde** [comment]
I'm facing the same problem. The last working version of spring-boot-autoconfigure for me was 1.2.1

**Comment by jeromedoucet** [comment]
@ wilkinsona,

I only downgrade the auto-configure module.
spring-boot, spring-boot-starter, and all spring-boot-starter-* are 1.2.3

The app that illustrate the problem is here : https://github.com/spring-guides/gs-uploading-files , complete part : https://github.com/spring-guides/gs-uploading-files/tree/master/complete

Just package and deployed as a war in Weblogic 12.1.3

**Comment by mssoftde** [comment]
I did some additional tests and the only thing, besides downgrading spring-boot-autoconfigure, which worked for me was to create a MultipartConfigElement Bean, where I set the location. 
Setting it in application.properties didn't work.

**Comment by wilkinsona** [comment]
The root of the problem is that `OrderedCharacterEncodingFilter` is running after `HiddenHttpMethodFilter`. `HiddenHttpMethodFilter` triggers processing of the request body as it calls `getParameter` on the request. `OrderedCharacterEncodingFilter` then runs and sets the request's encoding. Setting the request's encoding after its body has been processed is bad and, on WebLogic, causes the request to lose track of all its multipart data.

A workaround is to disable the character encoding filter in `application.properties`:

```
spring.http.encoding.enabled: false
```

The cause of the problem appears to be the fix for #2148. This fix should have caused the character encoding filter to go first, but, on WebLogic at least, has made it go after `HiddenHttpMethodFilter`.

**Comment by mssoftde** [comment]
The workaround does not work when used with embedded Tomcat.
I suppose that Tomcat does not know where to store the uploaded files.

**Comment by wilkinsona** [comment]
This is a bug in WebLogic. The [javadoc](http://docs.oracle.com/javaee/7/api/javax/servlet/FilterRegistration.html#addMappingForUrlPatterns(java.util.EnumSet,%20boolean,%20java.lang.String...) states that "Filter mappings are matched in the order in which they were added" and we're adding the mappings in the correct order:

```
2015-05-06 14:26:40.247  INFO 82113 --- [ (self-tuning)'] o.s.b.c.embedded.FilterRegistrationBean  : Mapping filter: 'characterEncodingFilter' to: [/*]
2015-05-06 14:26:40.247  INFO 82113 --- [ (self-tuning)'] o.s.b.c.embedded.FilterRegistrationBean  : Mapping filter: 'hiddenHttpMethodFilter' to: [/*]
```

WebLogic appears to invert this order such that the Filter that was added last is processed first. I'd recommend either using the workaround documented above or pursuing this with WebLogic support.

**Comment by wilkinsona** [comment]
@mssoftde There's no need for the workaround if you're using Tomcat. The problem is specific to WebLogic.

**Comment by mssoftde** [comment]
@wilkinsona I had the same error message after upgrading from Spring Boot 1.2.1 to 1.2.3 when using the embedded Tomcat.

**Comment by wilkinsona** [comment]
@mssoftde The problem originally reported here was resolved by downgrading spring-boot-autoconfigure to 1.2.0 which fits with the filter ordering being made explicit in 1.2.1. If 1.2.1 works for you and 1.2.3 does not then you're facing a different issue.

I've just tested the `gs-uploading-files` sample with Boot 1.2.3 and embedded Tomcat and its works. Please open a separate issue describing how to recreate the problem if you'd like us to investigate.

**Comment by tan9** [comment]
For people who are looking for solution for this issue, I wrote a dirty workaround by wrapping `ServetContext` and try to correct the behavoir on WebLogic:

https://gist.github.com/tan9/fcc312caf520bc39fea5fb1da00aedbd

**Comment by unpringatmes** [comment]
It seems Oracle published patch 20493506 that resolves the problem
